How does decoder work ?

Hello,

I am trying to use rosc from tokio. My first step is to decode and display incoming osc messages. I have made the server:

```rust
use rosc::decoder;
use rosc::{OscMessage, OscPacket};
use tokio::io::AsyncReadExt;
use tokio::net::TcpListener;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("start");
    let listener = TcpListener::bind("0.0.0.0:9000").await?;
    println!("listening");
    loop {
        match listener.accept().await {
            Ok((mut socket, addr)) => {
                println!("new client: {:?}", addr);

                println!("ready");
                tokio::spawn(async move {
                    let mut buf = [0; 1024];

                    // In a loop, read data from the socket and write the data back.
                    loop {
                        let _ = match socket.read(&mut buf).await {
                            Ok(n) if n == 0 => {
                                eprintln!("end tran");
                                return;
                            }
                            Ok(n) => {
                                println!("{:?}", &buf[..n]);
                                println!("read {} bytes", n);
                                match decoder::decode_tcp(&buf[..n]) {
                                    Ok((_, payload)) => {
                                        println!("{:?}", payload);
                                        return;
                                    }
                                    _ => {
                                    eprintln!("decoding error");
                                    return;
                                    }
                                }
                            }
                            Err(e) => {
                                eprintln!("failed to read from socket; err = {:?}", e);
                                return;
                            }
                        };
                    }
                });
            }
            Err(e) => eprintln!("could not get client: {:?}", e),
        }
    }
}
```

and a simple client inspired from documentation:

```rust
use rosc::encoder;
use rosc::{OscMessage, OscPacket, OscType};
use tokio::io::{AsyncReadExt, AsyncWriteExt};
use tokio::net::TcpStream;
use tokio::time::{sleep, Duration};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let addr = "127.0.0.1:9000";
    let mut stream = TcpStream::connect(addr).await?;
    let osc_packet = OscPacket::Message(OscMessage {
        addr: "/greet/me".to_string(),
        args: vec![OscType::String("hi!".to_string())],
    });
    if let Ok(buf) = encoder::encode(&osc_packet) {
        println!("send packet");
        println!("{:?}", buf);
        stream.write_all(&buf).await?;
        println!("transmitted");
    }
    Ok(())
}
 ```

When the server receives the packet it seems to fail decoding it, and put everything in the left first  tuple element:

```
([47, 103, 114, 101, 101, 116, 47, 109, 101, 0, 0, 0, 44, 115, 0, 0, 104, 105, 33, 0], None)
```

I have made a little program to see what happens:

```rust
use rosc::{encoder,decoder};
use rosc::{OscMessage, OscPacket, OscType};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let osc_packet = OscPacket::Message(OscMessage {
        addr: "/greet/me".to_string(),
        args: vec![OscType::String("hi!".to_string())],
    });
    if let Ok(buf) = encoder::encode(&osc_packet) {
        println!("{:?}", buf);
        let payload = decoder::decode_tcp(&buf)?;
        println!("{:?}", payload);
    }
    Ok(())
}
```
got this output:

```
mik@debian:~/hello$ cargo run  --example test_decode
    Finished dev [unoptimized + debuginfo] target(s) in 0.01s
     Running `target/debug/examples/test_decode`
[47, 103, 114, 101, 101, 116, 47, 109, 101, 0, 0, 0, 44, 115, 0, 0, 104, 105, 33, 0]
([47, 103, 114, 101, 101, 116, 47, 109, 101, 0, 0, 0, 44, 115, 0, 0, 104, 105, 33, 0], None)
mik@debian:~/hello$
```

I guess I missuse the api. I tried to find something to decode directly from a buffer in the API without succeed.

I tried to reproduce the issue with your example files, but the server is not compiling:

```sh
$ cargo build --bin server
   Compiling example v0.1.0 (/private/var/folders/mc/3ssj7_c51dv76d6ll79s5l3w0000gn/T/tmp.pmm1mYUl/example)
error[E0432]: unresolved import `tokio::io::AsyncReadExt`
 --> src/server/server.rs:3:5
  |
3 | use tokio::io::AsyncReadExt;
  |     ^^^^^^^^^^^------------
  |     |          |
  |     |          help: a similar name exists in the module: `AsyncRead`
  |     no `AsyncReadExt` in `io`

error[E0432]: unresolved import `tokio::net::TcpListener`
 --> src/server/server.rs:4:5
  |
4 | use tokio::net::TcpListener;
  |     ^^^^^^^^^^^^^^^^^^^^^^^ no `TcpListener` in `net`
  |
help: consider importing this struct instead
  |
4 | use std::net::TcpListener;
  |     ~~~~~~~~~~~~~~~~~~~~~

error[E0433]: failed to resolve: could not find `main` in `tokio`
 --> src/server/server.rs:6:10
  |
6 | #[tokio::main]
  |          ^^^^ could not find `main` in `tokio`

error[E0425]: cannot find function `spawn` in crate `tokio`
  --> src/server/server.rs:17:24
   |
17 |                 tokio::spawn(async move {
   |                        ^^^^^ not found in `tokio`
   |
help: consider importing this function
   |
1  | use std::thread::spawn;
   |
help: if you import `spawn`, refer to it directly
   |
17 -                 tokio::spawn(async move {
17 +                 spawn(async move {
   |

warning: unused imports: `OscMessage`, `OscPacket`
 --> src/server/server.rs:2:12
  |
2 | use rosc::{OscMessage, OscPacket};
  |            ^^^^^^^^^^  ^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

error[E0752]: `main` function is not allowed to be `async`
 --> src/server/server.rs:7:1
  |
7 | async fn main() -> Result<(), Box<dyn std::error::Error>> {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `main` function is not allowed to be `async`

Some errors have detailed explanations: E0425, E0432, E0433, E0752.
For more information about an error, try `rustc --explain E0425`.
warning: `example` (bin "server") generated 1 warning
error: could not compile `example` due to 5 previous errors; 1 warning emitted
﻿```

--- a/src/address.rs
+++ b/src/address.rs
@@ -59,7 +59,7 @@ impl Matcher {
     /// - `{foo,bar}` is an alternative, matching either `foo` or `bar`
     /// - everything else is matched literally
     ///
-    /// Refer to the OSC specification for details about address pattern matching: <osc-message-dispatching-and-pattern-matching>.
+    /// Refer to the OSC specification for details about address pattern matching: <https://opensoundcontrol.stanford.edu/spec-1_0.html#osc-message-dispatching-and-pattern-matching>.
     ///
     /// # Examples
     ///

--- a/src/decoder.rs
+++ b/src/decoder.rs
@@ -32,6 +32,15 @@ pub fn decode_udp(msg: &[u8]) -> Result<(&[u8], OscPacket), OscError> {
 
 /// Takes a bytes slice from a TCP stream (or any stream-based protocol) and returns the first OSC
 /// packet as well as a slice of the bytes remaining after the packet.
+///
+/// # Difference to `decode_udp`
+///
+/// For stream-based protocols, such as TCP, the [OSC specification][^1] requires the size of
+/// the first packet to be send as a 32-bit integer before the actual packet data.
+///
+/// [^1]: _In a stream-based protocol such as TCP, the stream should begin with an int32 giving the size of the first packet, followed by the contents of the first packet, followed by the size of the second packet, etc._
+///
+/// [OSC specification]: https://cnmat.org/OpenSoundControl/OSC-spec.html
 pub fn decode_tcp(msg: &[u8]) -> Result<(&[u8], Option<OscPacket>), OscError> {
     let (input, osc_packet_length) = match be_u32(msg) {
         Ok((i, o)) => (i, o),

--- a/src/types.rs
+++ b/src/types.rs
@@ -15,7 +15,7 @@ use crate::alloc::{
 };
 
 /// A time tag in OSC message consists of two 32-bit integers where the first one denotes the number of seconds since 1900-01-01 and the second the fractions of a second.
-/// For details on its semantics see http://opensoundcontrol.org/node/3/#timetags
+/// For details on its semantics see <http://opensoundcontrol.org/node/3/#timetags>
 ///
 /// # Examples
 ///

